### PR TITLE
Adds initial implementation of news

### PR DESCRIPTION
**Pull request by rycee**
See #52 for the overall issue.

### CODE REVIEW

**Comment by rycee** [comment]
@nonsequitur @uvNikita You have been most active recently, would you mind giving this branch a try? It works stably on my system but I believe your setups differ from mine.

Please note that this version of Home Manager cannot work with the previous version's modules so before switching to this version make a note of the full path to the `home-manager` command and use that command to switch back to the master version.

**Comment by uvNikita** [comment]
A super nice idea of notifying users of changes! 

Just tried to switch to `add/news` branch and run `home-manager switch`. Everything worked as it should: I got a notification, read news with `home-manager news` command, and didn't get any notifications on next `home-manager switch` run. Next run of `home-manager news` shows `[read]` label near the message date.

I've also created a new entry in my `home.nix` config, which also worked as expected:
```
  news.entries = [
    {
      time = "2017-09-01T11:56:28+00:00";
      message = "Test";
    }
  ];
```

Am I right that once someone will create a new module in home-manager, they should also create a news entry inside this new module with something like this?
```
  news.entries = [
    {
      time = "2017-09-01T11:56:28+00:00";
      message = "vim: new module";
    }
  ];
```

And when there are some changes in the module, the news entries for this module would look something like this?
```
  news.entries = [
    {
      time = "2017-09-01T12:56:28+00:00";
      condition = cfg.plugins != [];
      message = ''
        vim: 'plugins' parameter was renamed to 'pkgs'.
        
        The old parameter will be deprecated in the future,
        please change your configuration file accordingly.
      '';
    };
    {
      time = "2017-09-01T11:56:28+00:00";
      message = "vim: new module";
    }
  ];
```

As an idea, we could also run `notify-send` command if it is available to show desktop notification about new entries.

**Comment by rycee** [comment]
Thanks for the testing and comment! I'm glad it worked for you :-)

I haven't spent much thought on the best way to communicate through the news feature. But I think that it should be different from the Git commit message structure since there is no "code" context. That is, I think it should be formatted more like a prose text and fully expand any option path. With this in mind, I would suggest rewording your examples to something like
```
  news.entries = [
    {
      time = "2017-09-01T12:56:28+00:00";
      condition = cfg.plugins != [];
      message = ''
        The 'programs.vim.plugins' parameter has been
        renamed to 'programs.vim.pkgs'.

        The old parameter is deprecated and will be removed in the future,
        please change your configuration file accordingly.
      '';
    }

    {
      time = "2017-09-01T11:56:28+00:00";
      message = ''
        A new module, 'vim', is now available under the 'programs.vim' option path.
      '';
    }
  ];
```

Note, for the example with renaming `plugins` to `pkgs` I believe the correct action would be to remove the news entry when removing the option declaration, otherwise the condition would no longer be sensible.

Oddly enough, I haven't though of using the news feature for announcing new modules but now that you mentioned it I think it is an obvious use case :-)

**Comment by rycee** [comment]
Oh, forgot about the notify-send. Yes, that seems like a nice idea! I'll try adding that.

**Comment by rycee** [comment]
@Infinisil Sorry, forgot to invite you to try this out! If you have time and interest please give it a go but note the steps necessary to go back to the master branch: https://github.com/rycee/home-manager/pull/53#issuecomment-326554089

**Comment by uvNikita** [comment]
Agree, it makes sense to use a full path to the options in news entries.

I also thought if it necessary to have separate `condition` parameter, or maybe it is just enough to conditionally add records to the entries list (not sure this is the right way to do it in nix):
```
mkIf programs.vim.plugins != [] {
  news.entries = [{ time = ".."; text = "..." }];
}
```

Also, considering that the actual news feed depends on the current configuration, maybe it makes sense to and `home-manager news --all` flag to show all news independently of the conditions and current configuration. It will be some sort of a project changelog. In this case, the `condition` parameter seems to be necessary, because all entries should always be present in the news.entries list :)

In the current implementation it can also be a case when the user will make a switch, read the news, then will modify the configuration (let's say that the user will enable vim module and will use deprecated `programs.vim.plugins` parameter) and will switch again. No notification will be shown (e.g. about vim module) since this news will be considered as "read". Even more, such news will be marked as "read" in the feed (`home-manager news`). One way to fix it would be to have a set of "read" news somewhere, which would be expanded on each `home-manager news` invocation. But not sure if we even should try to fix it at this stage :)

**Comment by rycee** [comment]
Yeah, the reason for having the condition is that I wanted the possibility of collecting all news. Initially I actually had not just `[(un)read]` as message flag but `[(un)read,(ir)relevant]`. I eventually removed that to only show relevant news for now but would like to keep the condition to be able to re-introduce it in the future. The reason for removing it was that I couldn't think of an elegant way to handle it in the modules. For example, in the zsh module the natural way to add an entry felt to me like
```
    config = mkIf cfg.enable {
      news.entries = [
        {
          time = "2017-08-23T10:51:54+00:00";
          message = ''
            The

                programs.zsh.historySize

            option has been removed, please use

                programs.zsh.history.size

            instead.
          '';
        }
      ];

      …
    };
```
but then the entry will never show up in the news, not even as "irrelevant". So one would have to write something a bit more cumbersome like
```
    config = mkMerge [
      {
        news.entries = [
          {
            time = "2017-08-23T10:51:54+00:00";
            condition = cfg.enable;
            message = ''
              The

                  programs.zsh.historySize

              option has been removed, please use

                  programs.zsh.history.size

              instead.
            '';
          }
        ];
      }

      mkIf cfg.enable {
        …
      }
    ];
```
A possibility I've considered it to simply have a separate module that just holds all the news entries, unfortunately it would mean that the modules won't be as standalone as before…

And you are absolutely right that news can end up marked as read when they are actually unread. I've been a bit concerned about this as well and think that the entries really need to be tracked individually. My idea here has been to identify the entries either by requiring them to have an `id` field (holding an UUID or something) or use a hash of the entry content. I'm a bit concerned about performance here but maybe using the `id` field together with Bash's associative will be fast enough.

File: home-manager/home-manager [hunk 1]
```diff
@@ -100,6 +136,47 @@ function doListPackages() {
...
+        --unread)
+            ${PAGER:-less} "$newsFileUnread"
+            ;;
+        *)
+            echo "Unknown argument $1"
+            return 1
+    esac
+
+    local dataDir="${XDG_DATA_HOME:-$HOME/.local/share}/home-manager"
+    local timestampPath="$dataDir/news-timestamp"
```
Comment (nonsequitur) [inline home-manager/home-manager]: `dataDir` and `timestampPath` are defined twice with the same values. Maybe make them global?

File: home-manager/home-manager [hunk 2]
```diff
@@ -67,18 +66,55 @@ function doBuild() {
...
+        errorEcho
+    elif [[ "$newsDisplay" == "show" ]]; then
+        doShowNews --unread
+    else
+        errorEcho "Unknown 'news.display' setting '$newsDisplay'."
+    fi
+}
+
+function doBuild() {
+    presentNews
```
Comment (nonsequitur) [inline home-manager/home-manager]: I'd present the news after the build.
1. In this way, news are only shown when the build succeeds.
2. You're more likely to miss the news when it's followed by build output, especially in verbose mode.

File: home-manager/home-manager [hunk 3]
```diff
@@ -67,18 +66,55 @@ function doBuild() {
...
+
+    doBuildAttr -A activationPackage
 }
 
 function doSwitch() {
+    presentNews
```
Comment (nonsequitur) [inline home-manager/home-manager]: See my previous comment in `doBuild`.

File: home-manager/home-manager (hunk 2 above)
Comment (rycee) [reply home-manager/home-manager]: Fair enough, I think that makes more sense. I'll move this to the bottom.

File: home-manager/home-manager (hunk 1 above)
Comment (rycee) [reply home-manager/home-manager]: Yeah, I'll change it to avoid the code duplication.

**Comment by rycee** [comment]
@nonsequitur Thanks for the comments!

**Comment by rycee** [comment]
Pushed some commits now that

- adds and uses news entry ID to check for read/unread (instead of just a time stamp, which as previously mentioned is quite brittle)
- presents news at the end of build/switch rather than the beginning
- uses notify-send to notify the users (if the command is available in PATH, I didn't want to depend on it).

**Comment by nonsequitur** [comment]
I've added a few [additional fixes](https://github.com/rycee/home-manager/compare/add/news...nonsequitur:add/news-fixups).

**Comment by rycee** [comment]
@nonsequitur Thanks for the improvements! I've adapted them and added here.

**Comment by rycee** [comment]
I've now cleaned up the commits slightly and rebased into master. Closing this issue. Many thanks for the help, it really made this much better!
